quill-engine/src/main/scala/io/getquill/ast/Ast.scala

```diff
@@ -221,113 +219,118 @@ case class Aggregation(operator: AggregationOperator, ast: Ast) extends Query {
     }
 }
 
-case class Take(query: Ast, n: Ast) extends Query {
-  def quat           = query.quat
-  def bestQuat: Quat = query.bestQuat
+final case class Take(query: Ast, n: Ast) extends Query {
+  override def quat: Quat     = query.quat
+  override def bestQuat: Quat = query.bestQuat
 }
 
-case class Drop(query: Ast, n: Ast) extends Query {
-  def quat           = query.quat
-  def bestQuat: Quat = query.bestQuat
+final case class Drop(query: Ast, n: Ast) extends Query {
+  override def quat: Quat     = query.quat
+  override def bestQuat: Quat = query.bestQuat
 }
 
-case class Union(a: Ast, b: Ast) extends Query {
-  def quat           = a.quat
-  def bestQuat: Quat = a.bestQuat
+final case class Union(a: Ast, b: Ast) extends Query {
+  override def quat: Quat     = a.quat
+  override def bestQuat: Quat = a.bestQuat
 } // a and b quats should be same
 
-case class UnionAll(a: Ast, b: Ast) extends Query {
-  def quat           = a.quat
-  def bestQuat: Quat = a.bestQuat
+final case class UnionAll(a: Ast, b: Ast) extends Query {
+  override def quat: Quat     = a.quat
+  override def bestQuat: Quat = a.bestQuat
 } // a and b quats should be same
 
-case class Join(
+final case class Join(
   typ: JoinType,
   a: Ast,
   b: Ast,
   aliasA: Ident,
   aliasB: Ident,
   on: Ast
 ) extends Query {
-  def quat           = Quat.Tuple(a.quat, b.quat)
-  def bestQuat: Quat = Quat.Tuple(a.bestQuat, b.bestQuat)
+  override def quat: Quat     = Quat.Tuple(a.quat, b.quat)
+  override def bestQuat: Quat = Quat.Tuple(a.bestQuat, b.bestQuat)
 }
 
-case class FlatJoin(typ: JoinType, a: Ast, aliasA: Ident, on: Ast) extends Query {
-  def quat           = a.quat
-  def bestQuat: Quat = a.bestQuat
+final case class FlatJoin(typ: JoinType, a: Ast, aliasA: Ident, on: Ast) extends Query {
+  override def quat: Quat     = a.quat
+  override def bestQuat: Quat = a.bestQuat
 }
 
-case class Distinct(a: Ast) extends Query {
-  def quat           = a.quat
-  def bestQuat: Quat = a.bestQuat
+final case class Distinct(a: Ast) extends Query {
+  override def quat: Quat     = a.quat
+  override def bestQuat: Quat = a.bestQuat
 }
 
-case class DistinctOn(query: Ast, alias: Ident, body: Ast) extends Query {
-  def quat           = query.quat
-  def bestQuat: Quat = query.bestQuat
+final case class DistinctOn(query: Ast, alias: Ident, body: Ast) extends Query {
+  override def quat: Quat     = query.quat
+  override def bestQuat: Quat = query.bestQuat
 }
 
-case class Nested(a: Ast) extends Query {
-  def quat           = a.quat
-  def bestQuat: Quat = a.bestQuat
+final case class Nested(a: Ast) extends Query {
+  override def quat: Quat     = a.quat
+  override def bestQuat: Quat = a.bestQuat
 }
 
 //************************************************************
 
 final class Infix(val parts: List[String], val params: List[Ast], val pure: Boolean, val transparent: Boolean)(
   theQuat: => Quat
 ) extends Ast {
-  def quat: Quat     = theQuat
```

Comment:
A few `theQuat` calls weren't "cached" and were potentially re-computed many times